src/btf.cpp

```diff
@@ -0,0 +1,200 @@
+#include <iostream>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/utsname.h>
+#include <string.h>
+#include "btf.h"
+#include "types.h"
+#include "bpftrace.h"
+
+#ifdef HAVE_LIBBPF_BTF_DUMP
+#include <linux/bpf.h>
+#include <linux/btf.h>
+#include <bpf/btf.h>
+#include <bpf/libbpf.h>
+
+#define MAX_ERRNO 4095
+#define IS_ERR_VALUE(x) ((x) >= (unsigned long)-MAX_ERRNO)
+
+namespace bpftrace {
+
+static inline long PTR_ERR(const void *ptr)
+{
+  return (long) ptr;
+}
+
+static inline bool IS_ERR(const void *ptr)
+{
+  return IS_ERR_VALUE((unsigned long) ptr);
+}
+
+static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
+
+static unsigned char *get_data(const char *file, ssize_t *sizep)
+{
+  ssize_t ret;
+  int fd;
+
+  fd = open(file, O_RDONLY);
+  if (fd < 0)
+    return NULL;
+
+  struct stat st;
+
+  if (fstat(fd, &st))
+  {
+    close(fd);
+    return NULL;
+  }
+
+  unsigned char *data;
+  unsigned int size;
+
+  size = st.st_size;
+
+  data = (unsigned char *) malloc(size);
+  if (!data)
+  {
+    close(fd);
+    return NULL;
+  }
+
+  ret = read(fd, data, size);
+  if (ret <= 0)
+  {
+    close(fd);
+    free(data);
+    return NULL;
+  }
+
+  close(fd);
+
+  *sizep = size;
+  return data;
+}
+
+void BTF::init_test(char *file)
+{
+  unsigned char *data;
+  ssize_t size;
+
+  data = get_data(file, &size);
+  if (data == NULL)
+  {
+    std::cerr << "BTF: failed to read data from: " << file << std::endl;
+    btf = NULL;
+    return;
+  }
+
+  btf = btf__new(data, (__u32) size);
+  if (IS_ERR(btf))
```

Comment:
the way it's supposed to be done with libbpf without (re-)defining your own IS_ERR/PTR_ERR, etc is to use libbpf_get_error() function. Like this:

```
int err;

btf = btf__new(...);
err = libbpf_get_error(btf);
if (err) {
    // error handling
}
```

Reply:
Then you can just drop MAX_ERRNO, IS_ERR, PTR_ERR, IS_ERR_VALUE definitions.

Reply:
ok, will change, thanks

Reply:
new version pushed with review comments changes